src/infrasys/h5_time_series_storage.py

```diff
@@ -115,28 +124,129 @@ def add_time_series(
+
+        Raises
+        ------
+        NotImplementedError
+            If no handler is implemented for the given metadata type
+        """
+        msg = f"Bug: need to implement add_time_series for {type(metadata)}"
+        raise NotImplementedError(msg)
+
+    @_add_time_series_dispatch.register(SingleTimeSeriesMetadata)
+    def _(
+        self,
+        metadata: SingleTimeSeriesMetadata,
+        time_series: SingleTimeSeries,
+        context: Any = None,
+        **kwargs: Any,
+    ) -> None:
+        """Store a SingleTimeSeries array.
+
+        Parameters
+        ----------
+        metadata : :class:`~infrasys.time_series_models.SingleTimeSeriesMetadata`
+            Metadata for the single time series
+        time_series : :class:`~infrasys.time_series_models.SingleTimeSeries`
+            Single time series data to store
+        context : Any
+            HDF5 file handle
+
+        See Also
+        --------
+        :meth:`add_time_series` : Public method for adding time series.
+        """
+        assert context is not None
+
+        root = context[self.HDF5_TS_ROOT_PATH]
+        uuid = str(metadata.time_series_uuid)
+
+        if uuid not in root:
+            group = root.create_group(uuid)
+
+            group.create_dataset(
+                "data", data=time_series.data_array, compression="gzip", compression_opts=5
+            )
+
+            group.attrs["type"] = metadata.type
+            group.attrs["initial_timestamp"] = metadata.initial_timestamp.isoformat()
+            group.attrs["resolution"] = metadata.resolution.total_seconds()
+
+    @_add_time_series_dispatch.register(DeterministicMetadata)
+    def _(
+        self,
+        metadata: DeterministicMetadata,
+        time_series: DeterministicTimeSeries,
+        context: Any = None,
+        **kwargs: Any,
+    ) -> None:
+        """Store a DeterministicTimeSeries array.
+
+        Parameters
+        ----------
+        metadata : :class:`~infrasys.time_series_models.DeterministicMetadata`
+            Metadata for the deterministic time series
+        time_series : :class:`~infrasys.time_series_models.DeterministicTimeSeries`
+            Deterministic time series data to store
+        context : Any
+            HDF5 file handle
+
+        See Also
+        --------
+        :meth:`add_time_series` : Public method for adding time series.
         """
-        assert isinstance(time_series, SingleTimeSeries)
-        assert isinstance(metadata, SingleTimeSeriesMetadata)
+        assert context is not None
+
         root = context[self.HDF5_TS_ROOT_PATH]
         uuid = str(metadata.time_series_uuid)
 
         if uuid not in root:
             group = root.create_group(uuid)
 
             group.create_dataset(
-                "data", data=time_series.data, compression="gzip", compression_opts=5
+                "data", data=time_series.data_array, compression="gzip", compression_opts=5
```

Comment:
Should there be an option to not use compression? If you ever benchmark against Arrow, this will make a difference, especially if the test does slicing and even more if the code has to span chunks.

Reply:
Sure. I changed the API such that we pass `compression_level`.